micro-rdk/src/common/data_store.rs

```diff
@@ -0,0 +1,514 @@
+                        &mut DATA_STORE[start_idx..end_idx],
+                        0,
+                        0,
+                    ));
+                }
+            }
+            return Ok(Self {
+                buffers,
+                collector_keys,
+            });
+        }
+        Err(DataStoreError::DataStoreInitialized)
+    }
+
+    fn get_index_for_collector(
+        &self,
+        collector_key: &ResourceMethodKey,
+    ) -> Result<usize, DataStoreError> {
+        self.collector_keys
+            .iter()
+            .position(|key| key == collector_key)
+            .ok_or(DataStoreError::UnknownCollectorKey(collector_key.clone()))
+    }
+}
+
+impl DataStore for StaticMemoryDataStore {
+    fn store_message(
+        &mut self,
+        collector_key: ResourceMethodKey,
+        message: SensorData,
+        overwrite_old_data: bool,
+    ) -> Result<(), DataStoreError> {
+        let buffer_index = self.get_index_for_collector(&collector_key)?;
+        let buffer = &mut self.buffers[buffer_index];
+        let encode_len = message.encoded_len();
+        let total_encode_len = length_delimiter_len(encode_len) + encode_len;
+        if total_encode_len > buffer.vacant_len() {
+            if overwrite_old_data {
+                let mut cons = unsafe {
+                    Consumer::new(buffer as &LocalRb<u8, &'static mut [MaybeUninit<u8>]>)
+                };
+                // we'll skip messages until we can fit the current one
+                while total_encode_len > buffer.vacant_len() {
+                    let encoded_len = leb128::read::unsigned(&mut cons)? as usize;
+                    cons.skip(encoded_len);
+                }
+            } else {
+                return Err(DataStoreError::DataBufferFull(collector_key, message));
+            }
+        }
+        unsafe {
+            let mut prod = Producer::new(buffer as &LocalRb<u8, &'static mut [MaybeUninit<u8>]>);
+            let (left, right) = prod.free_space_as_slices();
+            let mut chained = BufMut::chain_mut(left, right);
+            message.encode_length_delimited(&mut chained)?;
+            prod.advance(total_encode_len);
+        }
+
+        Ok(())
+    }
+    fn read_next_message(
+        &mut self,
+        collector_key: ResourceMethodKey,
+    ) -> Result<BytesMut, DataStoreError> {
+        let buffer_index = self.get_index_for_collector(&collector_key)?;
+        let buffer = &mut self.buffers[buffer_index];
+        if buffer.iter().peekable().peek().is_none() {
+            return Ok(BytesMut::with_capacity(0));
+        }
+        let mut cons =
+            unsafe { Consumer::new(buffer as &LocalRb<u8, &'static mut [MaybeUninit<u8>]>) };
+        let encoded_len = leb128::read::unsigned(&mut cons)? as usize;
+        let mut msg_bytes = BytesMut::with_capacity(encoded_len);
+        unsafe {
+            msg_bytes.set_len(encoded_len);
+        }
+        cons.pop_slice(&mut msg_bytes);
```

Comment:
Is there a way to get some sort of Writer on BytesMut such that the `unsafe set_len` isn't needed?

Reply:
Still interested in this, btw, if you have ideas.

Reply:
Couldn't find a way to do this that didn't involve allocating an extra copy

Reply:
OK, thanks for checking.